test/phoenix/integration/socket_driver_test.exs

```diff
@@ -0,0 +1,301 @@
+    use Phoenix.Endpoint, otp_app: :phoenix
+  end
+
+  defmodule MyChannel do
+    use Phoenix.Channel
+
+    def join(_topic, message, socket) do
+      if message[:alias], do: Process.register(self, message[:alias])
+
+      if message[:error] == true do
+        {:error, %{foo: :bar}}
+      else
+        send(self, {:after_join, message})
+        {:ok, socket}
+      end
+    end
+
+    def handle_info({:after_join, message}, socket) do
+      broadcast socket, "user_entered", %{user: message["user"]}
+      push socket, "joined", Map.merge(%{status: "connected"}, socket.assigns)
+      {:noreply, socket}
+    end
+
+    def handle_in("new_msg", message, socket) do
+      broadcast! socket, "new_msg", message
+      {:noreply, socket}
+    end
+
+    def handle_in("ping", message, socket) do
+      {:reply, {:ok, %{pong: message}}, socket}
+    end
+
+    def handle_in("stop", _message, socket) do
+      {:stop, :normal, socket}
+    end
+
+    def terminate(_reason, socket) do
+      push socket, "you_left", %{message: "bye!"}
+      :ok
+    end
+  end
+
+  defmodule MySocket do
+    use Phoenix.Socket
+
+    transport :transport, MyTransport
+
+    channel "rooms:*", MyChannel
+
+    def connect(params, socket) do
+      Logger.disable(self())
+
+      if params[:error] == true do
+        :error
+      else
+        {:ok, assign(socket, :socket_id, params[:socket_id])}
+      end
+    end
+
+    def id(socket), do: socket.assigns[:socket_id]
+  end
+
+
+  setup_all do
+    capture_log fn -> Endpoint.start_link() end
+    :ok
+  end
+
+  test "provides the protocol version" do
+    assert Version.match?(Driver.protocol_version(), "~> 1.0")
+  end
+
+  test "successful initialization" do
+    assert {:ok, _state} = init_driver()
+  end
+
+  test "invalid vsn" do
+    capture_log(fn -> assert :error = init_driver(%{"vsn" => "0.1.0"}) end)
+  end
+
+  test "initialization error" do
+    assert :error = init_driver(%{error: true})
+  end
+
+  test "join" do
+    {:ok, state} = init_driver()
+
+    assert {:ok, [%Reply{status: :ok, topic: "rooms:lobby"}], state} = join(state, "rooms:lobby")
+    assert_receive {:channel_push, %Message{topic: "rooms:lobby", event: "user_entered"} = out_msg} = msg
+    assert {:ok, [^out_msg], _state} = Driver.handle_info(msg, state)
+  end
+
+  test "double join" do
+    {:ok, state} = init_driver()
+    {:ok, _, state} = join(state, "rooms:lobby")
+
+    log = capture_log fn ->
+      assert  {:ok, [], _state} = join(state, "rooms:lobby")
+      assert_receive(
+        {
+          :channel_push,
+          %Reply{topic: "rooms:lobby", payload: %{reason: "already joined"}} = out_msg
+        } = msg
+      )
+      assert {:ok, [^out_msg], _state} = Driver.handle_info(msg, state)
+    end
+
+    assert log =~ "received join event with topic \"rooms:lobby\" but channel already joined"
+  end
+
+  test "joining an unmatched channel" do
+    {:ok, state} = init_driver()
+
+    assert  {
```

Comment:
It feels this indention is uncommon on Phoenix source. My proposal is to break it in two parts:

```elixir
assert {:error, :unmatched_topic, [%Reply{} = reply], _state} = join(...)
assert %{topic: ....} = reply
```

We should fix the other examples through the test source.